runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/PortablePipelineJarCreator.java

```diff
@@ -0,0 +1,289 @@
+    manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+    boolean classHasMainMethod = false;
+    try {
+      mainClass.getMethod("main", String[].class);
+      classHasMainMethod = true;
+    } catch (NoSuchMethodException e) {
+      LOG.warn("No Main-Class will be set in jar because {} lacks a main method.", mainClass);
+    }
+    if (classHasMainMethod) {
+      manifest.getMainAttributes().put(Name.MAIN_CLASS, mainClass.getName());
+    }
+    return manifest;
+  }
+
+  /** Copy resources from {@code classLoader} to {@code outputStream}. */
+  private static void writeClassPathResources(ClassLoader classLoader, JarOutputStream outputStream)
+      throws IOException {
+    List<String> classPathResources =
+        PipelineResources.detectClassPathResourcesToStage(classLoader);
+    Preconditions.checkArgument(
+        classPathResources.size() == 1, "Expected exactly one jar on " + classLoader.toString());
+    copyResourcesFromJar(new JarFile(classPathResources.get(0)), outputStream);
+  }
+
+  /** Copy resources from {@code inputJar} to {@code outputStream}. */
+  @VisibleForTesting
+  protected static void copyResourcesFromJar(JarFile inputJar, JarOutputStream outputStream)
+      throws IOException {
+    Enumeration<JarEntry> inputJarEntries = inputJar.entries();
+    // The zip spec allows multiple files with the same name; the Java zip libraries do not.
+    // Keep track of the files we've already written to filter out duplicates.
+    // Also, ignore the old manifest; we want to write our own.
+    Set<String> previousEntryNames = new HashSet<>(ImmutableList.of(JarFile.MANIFEST_NAME));
+    while (inputJarEntries.hasMoreElements()) {
+      JarEntry inputJarEntry = inputJarEntries.nextElement();
+      InputStream inputStream = inputJar.getInputStream(inputJarEntry);
+      String entryName = inputJarEntry.getName();
+      if (previousEntryNames.contains(entryName)) {
+        LOG.debug("Skipping duplicated file {}", entryName);
+      } else {
+        JarEntry outputJarEntry = new JarEntry(inputJarEntry);
+        outputStream.putNextEntry(outputJarEntry);
+        LOG.trace("Copying jar entry {}", inputJarEntry);
+        IOUtils.copy(inputStream, outputStream);
+        previousEntryNames.add(entryName);
+      }
+    }
+  }
+
+  private static void writePipeline(Pipeline pipeline, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    pipeline.writeTo(outputStream);
+  }
+
+  private static void writePipelineOptions(Struct optionsProto, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_OPTIONS_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    optionsProto.writeTo(outputStream);
+  }
+
+  @VisibleForTesting
+  interface ArtifactRetriever {
```

Comment:
Can we directly use ArtifactRetrievalServiceBlockingStub and mock it where needed?

Reply:
ArtifactRetrievalServiceBlockingStub is a final class, and the version of Mockito we are using does not seem to support mocking final classes.